examples/websocket_async_echo_server.hpp

```diff
@@ -0,0 +1,375 @@
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#ifndef BEAST_WEBSOCKET_ASYNC_ECHO_PEER_H_INCLUDED
+#define BEAST_WEBSOCKET_ASYNC_ECHO_PEER_H_INCLUDED
+
+#include <beast/core/placeholders.hpp>
+#include <beast/core/streambuf.hpp>
+#include <beast/websocket/stream.hpp>
+#include <boost/lexical_cast.hpp>
+#include <boost/optional.hpp>
+#include <atomic>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <ostream>
+#include <string>
+#include <thread>
+#include <type_traits>
+#include <typeindex>
+#include <unordered_map>
+#include <utility>
+
+namespace websocket {
+
+/** Asynchronous WebSocket echo client/server
+*/
+class async_echo_server
+{
+public:
+    using error_code = beast::error_code;
+    using address_type = boost::asio::ip::address;
+    using socket_type = boost::asio::ip::tcp::socket;
+    using endpoint_type = boost::asio::ip::tcp::endpoint;
+
+private:
+    struct identity
+    {
+        template<class Body, class Fields>
+        void
+        operator()(beast::http::message<
+            true, Body, Fields>& req) const
+        {
+            req.fields.replace("User-Agent", "async_echo_client");
+        }
+
+        template<class Body, class Fields>
+        void
+        operator()(beast::http::message<
+            false, Body, Fields>& resp) const
+        {
+            resp.fields.replace("Server", "async_echo_server");
+        }
+    };
+
+    /** A container of type-erased option setters.
+    */
+    template<class NextLayer>
+    class options_set
+    {
+        // workaround for std::function bug in msvc
+        struct callable
+        {
+            virtual ~callable() = default;
+            virtual void operator()(
+                beast::websocket::stream<NextLayer>&) = 0;
+        };
+
+        template<class T>
+        class callable_impl : public callable
+        {
+            T t_;
+
+        public:
+            template<class U>
+            callable_impl(U&& u)
+                : t_(std::forward<U>(u))
+            {
+            }
+
+            void
+            operator()(beast::websocket::stream<NextLayer>& ws)
+            {
+                t_(ws);
+            }
+        };
+
+        template<class Opt>
+        class lambda
+        {
+            Opt opt_;
+
+        public:
+            lambda(lambda&&) = default;
+            lambda(lambda const&) = default;
+
+            lambda(Opt const& opt)
+                : opt_(opt)
+            {
+            }
+
+            void
+            operator()(beast::websocket::stream<NextLayer>& ws) const
+            {
+                ws.set_option(opt_);
+            }
+        };
+
+        std::unordered_map<std::type_index,
+            std::unique_ptr<callable>> list_;
+
+    public:
+        template<class Opt>
+        void
+        set_option(Opt const& opt)
+        {
+            std::unique_ptr<callable> p;
```

Comment:
Beast can only use C++11. Are you suggesting an alternative piece of code? If so, I'd like to see it :)